src/Interpreters/InterpreterSystemQuery.cpp

```diff
@@ -905,12 +905,23 @@ StoragePtr InterpreterSystemQuery::tryRestartReplica(const StorageID & replica,
         throw Exception(ErrorCodes::ABORTED, "Database {} is being dropped or detached, will not restart replica {}",
                         backQuoteIfNeed(replica.getDatabaseName()), replica.getNameForLogs());
 
-    auto [database, table] = DatabaseCatalog::instance().tryGetDatabaseAndTable(replica, getContext());
+    std::optional<Exception> exception;
+    auto [database, table] = DatabaseCatalog::instance().getTableImpl(replica, getContext(), &exception);
     ASTPtr create_ast;
 
-    /// Detach actions
-    if (!table || !dynamic_cast<const StorageReplicatedMergeTree *>(table.get()))
+    if (!table)
+    {
+        if (throw_on_error)
+            throw Exception(*exception);
         return nullptr;
```

Comment:
Maybe we should log the error. Not sure if it would be useful, but restartReplicas() seems to ignore these errors.

Reply:
Indeed, good catch. Added.